:sparkles: Add extra root volume options

johnharris85 opened the pull request:
Signed-off-by: John Harris <joharris@vmware.com>

<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, minor or feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🏃 (:running:, other) -->

**What this PR does / why we need it**: 
Allows users to specify additional options for a machine's root volume.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1403 

**Notes**:
- If a user wants to specify a KMS key, they need to ensure the relevant user / role has use permissions on the key. Should this be documented somewhere? (if so, where?)
- If EncryptionKey is set, we set Encrypted to `true` (even if it was omitted / set to `false`). Thought this was a reasonable assumption, but open to changes there (error if key set but encrypted missing / `false`).

ncdc commented:
/ok-to-test
/assign

ncdc commented on api/v1alpha3/awsmachine_webhook.go:
```diff
@@ -43,6 +43,15 @@ var _ webhook.Validator = &AWSMachine{}
+	var allErrs field.ErrorList
+
+	if r.Spec.RootVolumeOptions != nil && r.Spec.RootVolumeOptions.Type == "io1" && r.Spec.RootVolumeOptions.IOPS == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("spec.rootVolumeOptions.iops"), "iops required if type is 'io1'"))
```
```suggestion
		allErrs = append(allErrs, field.Required(field.NewPath("spec", "rootVolumeOptions", "iops"), "iops required if type is 'io1'"))
```

ncdc commented on api/v1alpha3/types.go:
```diff
@@ -516,12 +516,35 @@ type Instance struct {
+	// +optional
+	IOPS int64 `json:"iops,omitempty"`
+
+	// Encrypted is whether the volume should be encrypted or not. If EncryptionKey is set it will be used, if not the default AWS key will be used.
```
I would move the details about EncryptionKey down to that field

johnharris85 commented:
/retest

johnharris85 commented:
@vincepri pointed out on Slack this should have some restore logic built in to make sure we don't miss these fields from annotations. Will add that and address Andy's comments later today.

vincepri commented:
Sounds good! Holding the PR until we have those in place

/hold

jzhoucliqr commented on pkg/cloud/services/ec2/instances.go:
```diff
@@ -358,19 +358,35 @@ func (s *Service) runInstance(role string, i *infrav1.Instance) (*infrav1.Instan
 	}
 
-	if i.RootDeviceSize != 0 {
+	if i.RootVolumeOptions != nil && i.RootVolumeOptions.Size != 0 {
```
so we'll ignore other options if Size==0 ?

johnharris85 replied:
If the size is 0 we don't create the disk so there's nothing for the options to apply to / be relevant to afaik?

jzhoucliqr replied:
so there is no default size? If so, shall we mark size as non-optional?

rudoi replied:
The default size from AWS is actually 8, if that info helps. This 0 check probably originally should have been a nil check with Size as a nil-able field. I forget if there was a reason why I didn't implement it that way the first time.

jzhoucliqr replied:
From ec2 sdk:  	
        // The size of the volume, in GiB.
	//
	// Default: If you're creating the volume from a snapshot and don't specify
	// a volume size, the default is the snapshot size.
	//
	// Constraints: 1-16384 for General Purpose SSD (gp2), 4-16384 for Provisioned
	// IOPS SSD (io1), 500-16384 for Throughput Optimized HDD (st1), 500-16384 for
	// Cold HDD (sc1), and 1-1024 for Magnetic (standard) volumes. If you specify
	// a snapshot, the volume size must be equal to or larger than the snapshot
	// size.
	VolumeSize *int64 `locationName:"volumeSize" type:"integer"`

johnharris85 replied:
Fine however we want to implement this, but if you're always going to get a disk maybe we should make it mandatory so there's not ambiguous behavior with the other options etc?

vincepri replied:
I'm fine having the struct and Size default to nil. Let's make sure to copy the default behavior from the SDK to our Size field.

johnharris85 replied:
Still not 100% clear on what our behavior should be though. If I specify 0 (currently, or nil after these changes) for the size, I'm going to get a volume with 8 (according to AWS default? Where does this come from? Want to avoid random hard-coded values or copying too much AWS logic). Should we still apply the other options (encryption etc? assuming yes). What's the difference (from a user POV) between size = 0 and size = nil? Does allowing 0 even make sense (what am I expecting if I put 0 for a size?)

jzhoucliqr replied:
these are good questions. just my understanding, please correct me if i'm wrong: size=0 and size=nil won't make much difference from user's PoV, both will create a volume with the size of the snapshot that backs the AMI. 

provide option to use default size is valuable, because the non-zero size specified must be >= snapshot size. it's quite common that we don't know the underline AMI snapshot size, so use default size is a safe option.

vincepri replied:
Actually, I'll defer to @detiber and @randomvariable for this, the most important part for me is to be consistent with the experience we're providing to our users

vincepri commented on api/v1alpha2/awsmachine_conversion.go:
```diff
@@ -19,19 +19,46 @@ package v1alpha2
+func restoreAWSMachineSpec(restored *infrav1alpha3.AWSMachineSpec, dst *infrav1alpha3.AWSMachineSpec) {
+	if restored.RootVolumeOptions != nil {
+		dst.RootVolumeOptions = restored.RootVolumeOptions
+	}
```
We can probably use `dst.RootVolumeOptions.DeepCopyInto(...)`

vincepri replied:
Also note that this would overwrite the Manual conversion for `RootDeviceSize` that we have in `Convert_v1alpha2_AWSMachineSpec_To_v1alpha3_AWSMachineSpec`

vincepri commented on api/v1alpha3/types.go:
```diff
@@ -516,12 +516,35 @@ type Instance struct {
-	// Specifies size (in Gi) of the root storage device
-	RootDeviceSize int64 `json:"rootDeviceSize,omitempty"`
+	// Configuration options for the root storage volume
+	RootVolumeOptions *RootVolumeOptions `json:"rootVolumeOptions,omitempty"`
```
```suggestion
	RootVolume *RootVolumeOptions `json:"rootVolume,omitempty"`
```

vincepri replied:
This also need `//+optional`

johnharris85 replied:
You want this renaming all over? s/RootVolumeOptions/RootVolume/g

vincepri replied:
Yeah, `Options` is redundant given that this field is already within `spec:`

vincepri commented on api/v1alpha3/types.go:
```diff
@@ -516,12 +516,35 @@ type Instance struct {
 }
+
+// RootVolumeOptions encapsulates the configuration options for the root volume
+type RootVolumeOptions struct {
```
```suggestion
type RootVolume struct {
```

johnharris85 commented:
Addressed most of @vincepri comments (except the [size question](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1498#pullrequestreview-351211540)) and rebased.

vincepri commented:
@johnharris85 Regarding https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1498, I think the only thing I'd add is a kubebuilder validation tag that checks that Size is greater than 0, and remove all the checks where we do `.Size != 0`. 

This makes sure that when a user sets the `rootVolume` struct, the `size` field should never be zero, if the struct is not-nil.

johnharris85 commented on pkg/cloud/services/ec2/instances.go:
```diff
@@ -564,6 +573,23 @@ func (s *Service) getImageRootDevice(imageID string) (*string, error) {
+
+	return output.Images[0].BlockDeviceMappings[0].Ebs.VolumeSize, nil
+}
+
```
This func is basically identical to the one above (to get the root device name), just returning a different attribute. Thoughts on adjusting the above one (which is only used once) to return size and name to save an AWS call?

detiber replied:
+1 to consolidating the calls. This can also be handled as a followup.

johnharris85 commented on config/crd/bases/infrastructure.cluster.x-k8s.io_awsmachinetemplates.yaml:
```diff
@@ -397,13 +398,17 @@ spec:
+                              device. Must be greater than the image root snapshot
+                              size or 8 (whichever is greater).
                             format: int64
+                            minimum: 8
```
I know we talked about 10 here @vincepri but 8 seemed like a more sensible guard given that it's the default.

johnharris85 commented:
Discussion w/ @vincepri on Slack around how to handle the size wrt defaults / validation -> https://app.slack.com/client/T09NY5SBT/CD6U2V71N/thread/CD6U2V71N-1580776492.150700.

Addressed all comments I think but left a couple questions in for feedback, can squash when ready.

johnharris85 commented:
Failure looks like generation out of date, but `make verify` succeeds on my machine...? (and `make generate` doesn't generate any changes?) Am I missing something?

ncdc commented:
@johnharris85 try to rebase on top of master

vincepri commented:
```
api/v1alpha3/awsmachine_webhook_test.go:27: File is not `gofmt`-ed with `-s` (gofmt)
		name       string
api/v1alpha2/awsmachine_conversion_test.go:154: File is not `gofmt`-ed with `-s` (gofmt)
}
```

@johnharris85 seems linting errors now

randomvariable commented:
Should probably queue this one after #1525 so we get the object fuzzer, and then this will then need rebasing I think.

johnharris85 commented:
OK rebased now #1525 landed. Fuzzer found a couple bugs which I also fixed.
- `AWSClusterStatus` conversion funcs were using the auto conversions for `Instance`, changed to using the manual ones (with the additional volume options)
- `AWSCluster` conversion func needed to restore the bastion if present (from annotation)

randomvariable commented:
/unhold
as we have the restore logic in place.

randomvariable commented:
Should merge cleanly via 3-way, but merge after #1535

ncdc commented:
@johnharris85 could you please rebase & resolve conflicts - thanks!

detiber commented:
/lgtm
